### PR TITLE
tls: override ALPN in transport socket options

### PR DESCRIPTION
**Pull request by yxue**
Signed-off-by: crazyxy <yxyan@google.com>

Description: Override ALPN in transport socket options
Risk Level: Low
Testing: Unit test
Docs Changes: N/A
Release Notes: N/A 
Fixes #Issue: Part of https://github.com/envoyproxy/envoy/issues/8197

### CODE REVIEW

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc [hunk 1]
```diff
@@ -3795,6 +3795,38 @@ TEST_P(SslSocketTest, OverrideRequestedServerNameWithoutSniInUpstreamTlsContext)
...
+      filter_chain->mutable_tls_context()->mutable_common_tls_context()->add_tls_certificates();
+  server_cert->mutable_certificate_chain()->set_filename(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_dns_cert.pem"));
+  server_cert->mutable_private_key()->set_filename(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/extensions/transport_sockets/tls/test_data/san_dns_key.pem"));
+  envoy::api::v2::auth::CommonTlsContext* server_ctx =
+      filter_chain->mutable_tls_context()->mutable_common_tls_context();
+
+  envoy::api::v2::auth::UpstreamTlsContext client;
+  // envoy::api::v2::auth::CommonTlsContext* client_ctx = client.mutable_common_tls_context();
```
Comment (lizan) [inline test/extensions/transport_sockets/tls/ssl_socket_test.cc]: nit: unused?

File: source/common/network/application_protocol.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+namespace Envoy {
+namespace Network {
+
+/**
+ * ALPN to set in the upstream connection. Filters can use this one to override the ALPN in TLS
+ * context.
+ */
+class ApplicationProtocols : public StreamInfo::FilterState::Object {
+public:
+  ApplicationProtocols(const std::vector<std::string>& application_protocols)
```
Comment (lizan) [inline source/common/network/application_protocol.h]: nit: explicit for constructor with one argument.

File: source/common/upstream/cluster_manager_impl.cc [hunk 3]
```diff
@@ -1326,14 +1333,15 @@ ClusterManagerPtr ProdClusterManagerFactory::clusterManagerFromProto(
...
 Http::ConnectionPool::InstancePtr ProdClusterManagerFactory::allocateConnPool(
     Event::Dispatcher& dispatcher, HostConstSharedPtr host, ResourcePriority priority,
-    Http::Protocol protocol, const Network::ConnectionSocket::OptionsSharedPtr& options) {
+    Http::Protocol protocol, const Network::ConnectionSocket::OptionsSharedPtr& options,
+    const Network::TransportSocketOptionsSharedPtr& transport_socket_options) {
   if (protocol == Http::Protocol::Http2 &&
       runtime_.snapshot().featureEnabled("upstream.use_http2", 100)) {
-    return Http::ConnectionPool::InstancePtr{
-        new Http::Http2::ProdConnPoolImpl(dispatcher, host, priority, options)};
+    return Http::ConnectionPool::InstancePtr{new Http::Http2::ProdConnPoolImpl(
```
Comment (lizan) [inline source/common/upstream/cluster_manager_impl.cc]: while you're here, change this to std::make_unique?

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc (hunk 1 above)
Comment (yxue) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc]: Removed.

File: source/common/upstream/cluster_manager_impl.cc (hunk 3 above)
Comment (yxue) [reply source/common/upstream/cluster_manager_impl.cc]: Updated.

File: source/common/network/application_protocol.h (hunk 2 above)
Comment (yxue) [reply source/common/network/application_protocol.h]: Updated.

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 4]
```diff
@@ -802,6 +802,16 @@ bssl::UniquePtr<SSL> ClientContextImpl::newSsl(const Network::TransportSocketOpt
     SSL_set_verify(ssl_con.get(), SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, nullptr);
   }
 
+  if (options && !options->applicationProtocolListOverride().empty()) {
+    std::vector<uint8_t> parsed_override_alpn =
+        parseAlpnProtocols(absl::StrJoin(options->applicationProtocolListOverride(), ","));
+    if (!parsed_override_alpn.empty()) {
+      int rc =
+          SSL_set_alpn_protos(ssl_con.get(), &parsed_override_alpn[0], parsed_override_alpn.size());
```
Comment (lizan) [inline source/extensions/transport_sockets/tls/context_impl.cc]: nit: parsed_override_alpn.data()

Comment (yxue) [reply source/extensions/transport_sockets/tls/context_impl.cc]: Updated. Thanks!

**Pull request merged**